### PR TITLE
feat(jwt) add ES256 signatures support

### LINKED ISSUE #1499: Support ES256 algorithm for JWT plugin
Currently the JWT implementation supports RS256. It would be nice if ECC signatures were supported. ES256 and ECC in general are more modern crypto algorithms, and the keys and signatures are much more compact.

### PR DESCRIPTION
**Pull request by vcsjones**
### Summary

Add support for ES256 signatures in the JWT plugin.
### Full changelog

*\* ES256 signature signing.
*\* ES256 signature verification.
### Issues resolved

Fix #1499 
### Details

This adds support for ES256 signatures for JWT in compliance with the JWT specification. The specification lists "ES256", or ECDSA256withSha256, as a "SHOULD" for support. ES256 is desirable because it offers smaller signatures and uses more modern ECC cryptography.

Since it appears that Kong specifically compiles again OpenSSL 1.0.1h, there should be no compatibility problems with an installation not supporting it.

ES256 signatures differ from RS256 signatures in a few ways. First, ECDSA has a non-deterministic value called "k" that is random. This means the signature changes every time the signature is computed, even if the data and key do not change. This means that deterministically testing the result of the signature is not possible in the unit tests, only that the signature correctly round trips.

I validated the signatures against other popular libraries from Node.js and .NET to verify that the signatures generated work on other platforms and meet the specifications.

### CODE REVIEW

**Comment by vcsjones** [comment]
It's also worth noting that this does not change the handling of RS or HS signatures, so they are safe in that regard.

**Comment by vcsjones** [comment]
Since this is my first time writing anything in Lua, I am sure I missed a few things. It looks like I am supposed to make some changes in the rockspec file? Anything else?

**Comment by vcsjones** [comment]
Some work is still needed on this. This should be marked as blocked.

**Comment by vcsjones** [comment]
This is complete, to my knowledge the failures are not caused by this PR.

File: kong/plugins/jwt/jwt_parser.lua [hunk 1]
```diff
@@ -23,7 +24,21 @@ local alg_sign = {
   ["HS256"] = function(data, key) return crypto.hmac.digest("sha256", data, key, true) end,
   --["HS384"] = function(data, key) return crypto.hmac.digest("sha384", data, key, true) end,
   --["HS512"] = function(data, key) return crypto.hmac.digest("sha512", data, key, true) end
-  ["RS256"] = function(data, key) return crypto.sign('sha256', data, crypto.pkey.from_pem(key, true)) end
+  ["RS256"] = function(data, key)
+    local pkeyPrivate = crypto.pkey.from_pem(key, true)
+    return crypto.sign('sha256', data, pkeyPrivate)
+  end,
```
Comment (Tieske) [inline kong/plugins/jwt/jwt_parser.lua]: the lines above do not have a functional change, please revert. This obfuscates the diff.

File: kong/plugins/jwt/asn_sequence.lua [hunk 2]
```diff
@@ -0,0 +1,82 @@
+local error = error
+
+local _M = {}
+_M.__index = _M
+
+function _M.create_simple_sequence(input)
+  if type(input) ~= "table" then error("Argument #1 must be a table", 2) end
+  local sortTable = {}
```
Comment (Tieske) [inline kong/plugins/jwt/asn_sequence.lua]: There is a lot of error checking done. I suggest dropping this. Just verify external inputs (from the plugin configuration, or an incoming request). And stop there.

Add testcases for bad input and make sure the proper return codes are generated without leaking sensitive information.

in general, for bad input/user errors; return `nil + errormsg`, for hard technical errors, use `error` or `assert` (the latter two will mostly generate 500's, whereas for the former you'd want 400's)

File: kong/plugins/jwt/asn_sequence.lua [hunk 3]
```diff
@@ -0,0 +1,82 @@
...
+  elseif length ~= #input-2 then error("Sequence's asn length does not match expected length") end
+  local seq = {}
+  local counter = 1
+  local position = 3
+  while true do
+    if position == #input+1 then break
+    elseif position > #input+1 then error("Sequence moved out of bounds.")
+    elseif counter > 0xFF then error("Sequence is too long") end
+    local chunk = string.sub(input, position)
+    if string.byte(chunk, 1) ~= 0x2 then error("Sequence did not contain integers") end
```
Comment (Tieske) [inline kong/plugins/jwt/asn_sequence.lua]: Code formatting; please stick to the formatting used in the rest of the codebase.

eg.
```lua
if position == #input+1 then 
  break
elseif position > #input+1 then 
  error("Sequence moved out of bounds.")
elseif counter > 0xFF then 
  error("Sequence is too long") 
end
```

**Comment by Tieske** [comment]
merged through #1920

**Pull request closed without merging**